File: app/src/test/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongFromUriRetrieverTest.java
package com.github.goldy1992.mp3player.service.library.content.retriever;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.media.MediaMetadataRetriever;
import android.net.Uri;

import androidx.test.platform.app.InstrumentationRegistry;

import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.github.goldy1992.mp3player.commons.MediaItemUtils;
import com.github.goldy1992.mp3player.service.library.MediaItemTypeIds;
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.Collections;

import static android.media.MediaMetadataRetriever.METADATA_KEY_ARTIST;
import static android.media.MediaMetadataRetriever.METADATA_KEY_DURATION;
import static android.media.MediaMetadataRetriever.METADATA_KEY_TITLE;
import static android.support.v4.media.MediaBrowserCompat.MediaItem;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class SongFromUriRetrieverTest {

    private SongFromUriRetriever songFromUriRetriever;

    @Mock
    private ContentResolver contentResolver;

    @Mock
    private SongResultsParser songResultsParser;

    @Mock
    private Uri testUri;

    @Mock
    private MediaMetadataRetriever mmr;

    private MediaItemTypeIds mediaItemTypeIds;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        this.mediaItemTypeIds = new MediaItemTypeIds();
        Context context = InstrumentationRegistry.getInstrumentation().getContext();
        this.songFromUriRetriever = new SongFromUriRetriever(context, contentResolver, songResultsParser, mmr, mediaItemTypeIds);
    }

    @Test
    public void testGetSongWithContentScheme() {
        when(testUri.getScheme()).thenReturn(ContentResolver.SCHEME_CONTENT);

        final byte[] expectedEmbeddedPic = new byte[1];
        when(mmr.getEmbeddedPicture()).thenReturn(expectedEmbeddedPic);

        final String expectedTitle = "TITLE";
        when(mmr.extractMetadata(METADATA_KEY_TITLE)).thenReturn(expectedTitle);

        final String expectedArtist = "ARTIST";
        when(mmr.extractMetadata(METADATA_KEY_ARTIST)).thenReturn(expectedArtist);

        final long expectedDuration = 1123L;
        when(mmr.extractMetadata(METADATA_KEY_DURATION)).thenReturn(String.valueOf(expectedDuration));


        MediaItem result = songFromUriRetriever.getSong(testUri);

        final byte[] actualEmbeddedPicture = MediaItemUtils.getAlbumArtImage(result);
        assertEquals(expectedEmbeddedPic, actualEmbeddedPicture);

        final String actualTitle = MediaItemUtils.getTitle(result);
        assertEquals(expectedTitle, actualTitle);

        final String actualArtist = MediaItemUtils.getArtist(result);
        assertEquals(expectedArtist, actualArtist);

        final long actualDuration = MediaItemUtils.getDuration(result);
        assertEquals(expectedDuration, actualDuration);

        final Uri actualMediaUri = MediaItemUtils.getMediaUri(result);
        assertEquals(testUri, actualMediaUri);
    }

    @Test
    public void testGetSongWithNonContentScheme() {

        final MediaItem expectedMediaItem = mock(MediaItem.class);
        Cursor cursor = mock(Cursor.class);
        when(contentResolver.query(any(), any(), any(), any(), any()))
                .thenReturn(cursor);

        final String id = mediaItemTypeIds.getId(MediaItemType.SONGS);
        when(songResultsParser.create(cursor, id)).thenReturn(Collections.singletonList(expectedMediaItem));
        when(testUri.getScheme()).thenReturn(ContentResolver.SCHEME_FILE);

        MediaItem result = songFromUriRetriever.getSong(testUri);

        assertEquals(expectedMediaItem, result);

    }

    @Test
    public void testUriWithNullScheme() {
        when(testUri.getScheme()).thenReturn(null);
        MediaItem result = songFromUriRetriever.getSong(testUri);
        assertNull(result);
    }
}

File: service/src/test/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongFromUriRetrieverTest.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import android.content.ContentResolver
import android.database.Cursor
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.support.v4.media.MediaBrowserCompat
import androidx.test.platform.app.InstrumentationRegistry
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.commons.MediaItemUtils.getAlbumArtImage
import com.github.goldy1992.mp3player.commons.MediaItemUtils.getArtist
import com.github.goldy1992.mp3player.commons.MediaItemUtils.getDuration
import com.github.goldy1992.mp3player.commons.MediaItemUtils.getMediaUri
import com.github.goldy1992.mp3player.commons.MediaItemUtils.getTitle
import com.github.goldy1992.mp3player.service.library.MediaItemTypeIds
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class SongFromUriRetrieverTest {
    private var songFromUriRetriever: SongFromUriRetriever? = null

    private val contentResolver: ContentResolver = mock<ContentResolver>()
    
    private val songResultsParser: SongResultsParser = mock<SongResultsParser>()

    private val testUri: Uri = mock<Uri>()

    private val mmr: MediaMetadataRetriever = mock<MediaMetadataRetriever>()
    private var mediaItemTypeIds: MediaItemTypeIds? = null
    @Before
    fun setup() {
        mediaItemTypeIds = MediaItemTypeIds()
        val context = InstrumentationRegistry.getInstrumentation().context
        songFromUriRetriever = SongFromUriRetriever(context, contentResolver!!, songResultsParser!!, mmr!!, mediaItemTypeIds!!)
    }

    @Test
    fun testGetSongWithContentScheme() {
        whenever(testUri!!.scheme).thenReturn(ContentResolver.SCHEME_CONTENT)
        val expectedEmbeddedPic = ByteArray(1)
        whenever(mmr!!.embeddedPicture).thenReturn(expectedEmbeddedPic)
        val expectedTitle = "TITLE"
        whenever(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_TITLE)).thenReturn(expectedTitle)
        val expectedArtist = "ARTIST"
        whenever(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ARTIST)).thenReturn(expectedArtist)
        val expectedDuration = 1123L
        whenever(mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)).thenReturn(expectedDuration.toString())
        val result = songFromUriRetriever!!.getSong(testUri)
        val actualEmbeddedPicture = getAlbumArtImage(result!!)
        Assert.assertEquals(expectedEmbeddedPic, actualEmbeddedPicture)
        val actualTitle = getTitle(result)
        Assert.assertEquals(expectedTitle, actualTitle)
        val actualArtist = getArtist(result)
        Assert.assertEquals(expectedArtist, actualArtist)
        val actualDuration = getDuration(result)
        Assert.assertEquals(expectedDuration, actualDuration)
        val actualMediaUri = getMediaUri(result)
        Assert.assertEquals(testUri, actualMediaUri)
    }

    @Test
    fun testGetSongWithNonContentScheme() {
        val expectedMediaItem = mock<MediaBrowserCompat.MediaItem>()
        val cursor = mock<Cursor>()
        whenever(contentResolver!!.query(any(), any(), eq(null), eq(null), eq(null)))
                .thenReturn(cursor)
        val id = mediaItemTypeIds!!.getId(MediaItemType.SONGS)
        whenever(songResultsParser!!.create(cursor, id!!)).thenReturn(listOf(expectedMediaItem))
        whenever(testUri!!.scheme).thenReturn(ContentResolver.SCHEME_FILE)
        val result = songFromUriRetriever!!.getSong(testUri)
        Assert.assertEquals(expectedMediaItem, result)
    }

    @Test
    fun testUriWithNullScheme() {
        whenever(testUri!!.scheme).thenReturn(null)
        val result = songFromUriRetriever!!.getSong(testUri)
        Assert.assertNull(result)
    }
}